### PR TITLE
FEAT: Spark tests

### PR DESCRIPTION
**Pull request by amroid**

### CODE REVIEW

File: ibis/spark/compiler.py [hunk 1]
```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
     pass
 
 
+def _is_inf(translator, expr):
+    arg, = expr.op().args
+    return '{} = {} or {} = {}'.format(
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Would this be cleaner as a rewrite rule?

File: ibis/spark/compiler.py [hunk 2]
```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
...
+
+
+def _cast(translator, expr):
+    op = expr.op()
+    arg, target_type = op.args
+    arg_formatted = translator.translate(arg)
+
+    if isinstance(arg, ir.CategoryValue) and target_type == dt.int32:
+        return arg_formatted
+    if isinstance(arg, ir.TemporalValue) and target_type == dt.int64:
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Do we have a test that hits this branch?

File: ibis/spark/compiler.py [hunk 3]
```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
...
+    op = expr.op()
+    left, right = op.args
+    casted_left = 'cast(to_timestamp({}) as double)'.format(
+        translator.translate(left)
+    )
+    casted_right = 'cast(to_timestamp({}) as double)'.format(
+        translator.translate(right)
+    )
+
+    return '{} - {}'.format(
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Parenthesize

File: ibis/spark/compiler.py [hunk 4]
```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
...
+
+    return 'array({})'.format(
+        ', '.join(translated_values)
+    )
+
+
+def _struct_like_format(func):
+    def formatter(translator, expr):
+        translated_values = [
+            ("'{}'".format(name), translator.translate(ibis.literal(val)))
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Use `'{!r}'` instead of embedding quotes.

File: ibis/spark/compiler.py [hunk 5]
```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
...
+        translator.translate(format_string)
+    )
+
+
+@rewrites(ops.DayOfWeekName)
+def spark_rewrites_day_of_week_name(expr):
+    arg = expr.op().arg
+    # TODO: if strftime gets rewritten to be compatible with ibis standard
+    # strftime, then 'EEEEE' needs to be rewritten to '%A'
+    return arg.strftime('EEEEE')
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Is this really how to specify the name of the day of week??

File: ibis/spark/compiler.py [hunk 6]
```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
...
+        if isinstance(right, (ir.TimestampScalar, ir.DateValue)):
+            formatted_right = 'cast({} as timestamp)'.format(formatted_right)
+
+        # Spark only supports date_add, date_sub, and add_months.
+        # date_add and date_sub are in days, add_months is in months.
+        # These ops set HH:MM:SS to zero, operating only on date.
+        #
+        # If we use the <timestamp> + INTERVAL <int> <unit> syntax,
+        # Spark only supports the second argument being a literal.
+        if func_name in ('date_add', 'date_sub'):
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Can we break this logic out into functions that encapsulate the various cases that are being handled here? It's hard to tell by reading the code which case is which.

File: ibis/impala/compiler.py [hunk 7]
```diff
@@ -631,10 +632,18 @@ def _string_literal_format(translator, expr):
...
+        formatted = repr(value)
+    else:
+        if math.isnan(value):
+            formatted_val = "'NaN'"
+        elif math.isinf(value):
+            if value > 0:
+                formatted_val = "'Infinity'"
+            else:
+                formatted_val = "'-Infinity'"
+        formatted = "CAST({} AS DOUBLE)".format(formatted_val)
```
Comment (cpcloud) [inline ibis/impala/compiler.py]: Can you use `"CAST({!r} AS DOUBLE)".format(fomatted_val)` and remove the embedded single quotes from nan, infinity, etc?

File: ibis/spark/compiler.py [hunk 8]
```diff
@@ -26,13 +68,365 @@ class SparkContext(ImpalaContext):
...
+    arg_formatted = translator.translate(arg)
+    try:
+        unit = _spark_timestamp_unit_names[unit]
+    except KeyError:
+        raise com.UnsupportedOperationError(
+            '{!r} unit is not supported in timestamp truncate'.format(unit)
+        )
+
+    if unit == 'DAY':
+        return "date(date_trunc('{}', {}))".format(unit, arg_formatted)
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Generally speaking prefer `{!r}` to `'{}'` inside string formatting.

File: ibis/spark/compiler.py [hunk 9]
```diff
@@ -26,13 +68,365 @@ class SparkContext(ImpalaContext):
...
+
+    arg_formatted = translator.translate(arg)
+    try:
+        unit = _spark_date_unit_names[unit]
+    except KeyError:
+        raise com.UnsupportedOperationError(
+            '{!r} unit is not supported in date truncate'.format(unit)
+        )
+
+    return "trunc({}, '{}')".format(arg_formatted, unit)
```
Comment (cpcloud) [inline ibis/spark/compiler.py]: Use `{!r}` here

File: ibis/spark/client.py [hunk 10]
```diff
@@ -180,6 +181,10 @@ def current_database(self):
     def _get_table_schema(self, table_name):
         return self.get_schema(table_name)
 
+    def _get_schema_using_query(self, query):
+        cur = self._execute(query, results=True)
+        return spark_dataframe_schema(cur.query)
```
[the remark this answers is not in the document]
Comment (amroid) [reply ibis/spark/client.py]: The name is a little overloaded here, `query` is a SQL string and `cur.query` is a `pyspark.sql.dataframe.DataFrame`.

**Comment by cpcloud** [comment]
ping @amroid what's the status here?

File: ibis/spark/compiler.py (hunk 5 above)
Comment (amroid) [reply ibis/spark/compiler.py]: Turns out just four Es suffices, major improvement

File: ibis/tests/all/test_temporal.py [hunk 11]
```diff
@@ -54,6 +55,8 @@ def test_timestamp_extract(backend, alltypes, df, attr):
         'Y',
         'M',
         'D',
+        # Spark truncation to week truncates to different days than Pandas
+        # Pandas backend is probably doing this wrong
```
[the remark this answers is not in the document]
Comment (amroid) [reply ibis/tests/all/test_temporal.py]: Made issue.

File: ibis/tests/all/test_temporal.py [hunk 12]
```diff
@@ -270,6 +276,8 @@ def test_interval_add_cast_column(backend, alltypes, df):
     ('ibis_pattern', 'pandas_pattern'), [('%Y%m%d', '%Y%m%d')]
 )
 @pytest.mark.xfail_unsupported
+# Spark takes Java SimpleDateFormat instead of strftime
```
[the remark this answers is not in the document]
Comment (amroid) [reply ibis/tests/all/test_temporal.py]: Made issue.

File: ibis/tests/backends.py [hunk 13]
```diff
@@ -522,3 +522,84 @@ def connect(data_directory: Path) -> ibis.client.Client:
...
+
+        client = ibis.spark.connect()
+
+        df_functional_alltypes = client._session.read.csv(
+            path=str(data_directory / 'functional_alltypes.csv'),
+            schema=pt.StructType([
+                pt.StructField('index', pt.IntegerType(), True),
+                pt.StructField('Unnamed: 0', pt.IntegerType(), True),
+                pt.StructField('id', pt.IntegerType(), True),
+                # cast below, Spark can't read 0/1 as bool
```
[the remark this answers is not in the document]
Comment (amroid) [reply ibis/tests/backends.py]: Made issue.

File: ibis/spark/api.py [hunk 14]
```diff
@@ -8,5 +9,6 @@ def connect(**kwargs):
     https://spark.apache.org/docs/latest/api/python/_modules/pyspark/context.html#SparkContext
     """
     client = SparkClient(**kwargs)
+    client._session.conf.set('spark.sql.session.timeZone', 'UTC')
```
[the remark this answers is not in the document]
Comment (amroid) [reply ibis/spark/api.py]: Spark internally stores timestamps as UTC values, and timestamp data that is brought in without a specified time zone is converted as local time to UTC with microsecond resolution.

https://spark.apache.org/docs/latest/sql-pyspark-pandas-with-arrow.html#timestamp-with-time-zone-semantics

File: ibis/spark/compiler.py (hunk 6 above)
Comment (amroid) [reply ibis/spark/compiler.py]: Moving to new branch.

File: ibis/tests/backends.py [hunk 15]
```diff
@@ -522,3 +522,84 @@ def connect(data_directory: Path) -> ibis.client.Client:
...
+                pt.StructField('playerID', pt.StringType(), True),
+                pt.StructField('awardID', pt.StringType(), True),
+                pt.StructField('yearID', pt.IntegerType(), True),
+                pt.StructField('lgID', pt.StringType(), True),
+                pt.StructField('tie', pt.StringType(), True),
+                pt.StructField('notes', pt.StringType(), True),
+            ]),
+            header=True,
+        )
+        df_awards_players.createOrReplaceTempView('awards_players')
```
[the remark this answers is not in the document]
Comment (amroid) [reply ibis/tests/backends.py]: The old `pyspark.sql.SQLContext` class, which has been replaced by `pyspark.sql.SparkSession`, has a method `registerDataFrameAsTable` that just calls `df.createOrReplaceTempView`.

The `pyspark.sql.Catalog` class has a `createTable` method, but it looks like this writes the data externally.

File: ibis/spark/compiler.py (hunk 2 above)
Comment (amroid) [reply ibis/spark/compiler.py]: Created issue.

File: ibis/spark/compiler.py [hunk 16]
```diff
@@ -26,13 +61,289 @@ class SparkContext(ImpalaContext):
...
+@compiles(ops.DayOfWeekName)
+def spark_compiles_day_of_week_name(translator, expr):
+    arg = expr.op().arg
+    return 'date_format({}, {!r})'.format(translator.translate(arg), 'EEEE')
+
+
+@rewrites(ops.IsInf)
+def spark_rewrites_is_inf(expr):
+    arg = expr.op().arg
+    return (arg == ibis.literal(math.inf, type='double')) | \
```
[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/spark/compiler.py]: `type='double'` isn't necessary here.

File: ibis/tests/all/conftest.py [hunk 17]
```diff
@@ -120,7 +120,9 @@ def pytest_pyfunc_call(pyfuncitem):
 
 
 @pytest.fixture(params=params_backend, scope='session')
-def backend(request, data_directory):
+def backend(request, data_directory, spark_client_testing):
+    if request.param is Spark:
+        Spark.client_testing = request.getfixturevalue('spark_client_testing')
```
Comment (cpcloud) [inline ibis/tests/all/conftest.py]: Let's understand why `Spark.client_testing = spark_client_testing` doesn't work.

Comment (amroid) [reply ibis/tests/all/conftest.py]: Works!

**Pull request merged**